qa/rpc-tests/randompruning.py

```diff
@@ -0,0 +1,153 @@
+
+        # Now interconnect the nodes
+        connect_nodes_full(self.nodes)
+        # Let the framework know if the network is fully connected.
+        # If not, the framework assumes this partition: (0,1) and (2,3)
+        # For more complex partitions, you can't use the self.sync* member functions
+        self.is_network_split=False
+        self.sync_blocks()
+
+    def run_test (self):
+        # generate enough blocks so that nodes[0] has a balance
+        # mine some number of blocks higher than the minimum we would be pruning
+        self.sync_blocks()
+        blocks_mined = []
+        for i in range(101):
+            blockhash = self.nodes[0].generate(1)[0]
+            # dont add blocks above the min blocks to keep line
+            if (i < 80):
+                blocks_mined.append(blockhash)
+        self.sync_blocks()
+
+        #get the max possible value of our hashmask
+        uint64_t_max = (2**64) - 1
+        # calculate 1% which
+        threshold_percent = uint64_t_max / 100
+        blockchaininfo = self.nodes[1].getblockchaininfo()
+        hashMask64 = blockchaininfo["pruneHashMask"]
+        threshold = blockchaininfo["hashMaskThreshold"]
+        # calculate the normalize threshold for the node
+        normalized_threshold = threshold * threshold_percent
+
+        # our threshold should be 10
+        assert_equal({'prune.hashMaskThreshold': 10}, self.nodes[1].get("prune.hashMaskThreshold"))
+
+        # check the blocks mined prior to determine if we are keeping the blocks we expect to have
+        # and pruning the ones we expect to prune
+
+        # kept does nothing in the actual test but it was/is useful for debugging if the test fails
+        kept = []
+        for block in blocks_mined:
+            assert_equal(len(block), 64)
+            low64block = GetLow64(block)
+            # if the 64LSB of the blockhash is equal to or above our threshold we should have pruned it
+            # assert this is the case
+            valxmask = int(low64block, 16) ^ hashMask64
+            if valxmask >= normalized_threshold:
+                assert_raises_rpc_error(-32603, "Block not available (pruned data)", self.nodes[1].getblock, block)
+            else:
+                kept.append(self.nodes[1].getblock(block)['height'])
+
+        print("checking if we can raise the threshold higher than the current one... (should not be able to)")
+        # we should get an error raising the threshold higher than 10% since that was its last value
+        assert_raises_rpc_error(-1, "current hashMaskThreshold (10) is smaller than your proposed new threshold (70)",
+        self.nodes[1].set, "prune.hashMaskThreshold=70")
+        # our threshold should still be 10
+        print("checking to make sure nothing has changed...")
+        assert_equal({'prune.hashMaskThreshold': 10}, self.nodes[1].get("prune.hashMaskThreshold"))
```

Comment:
Having only 100 divisions seems too few, especially to add a feature where we automatically reduce the threshold periodically to keep the blockchain size < a configured size in GB.  This feature is what people really want since they can set it to whatever is appropriate based on their disk size.

We could expose the raw threshold number, we could accept floating point numbers in the CLI, or we could used fixed point -- that is, the threshold is a percent * 1e6  (i.e. 1% -> prune.hashMaskThreshold=1000000)